File: dfs&bfs/BOJ-15651.py
import sys
n, m = map(int, sys.stdin.readline().split(' '))

selected = [0 for _ in range(m)]

def recur(k):
    if k == m:
        for x in selected:
            print(x, end = ' ')
        print()
    else:
        for cand in range(1, n + 1):
            selected[k] = cand
            recur(k + 1)

recur(0)

File: brute-force/BOJ-15651.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
)

var (
	sc       *bufio.Scanner
	wr       *bufio.Writer
	selected = make([]int, 0)
)

func init() {
	sc = bufio.NewScanner(os.Stdin)
	wr = bufio.NewWriter(os.Stdout)
	sc.Split(bufio.ScanWords)
}

func main() {
	sc.Scan()
	n, _ := strconv.Atoi(sc.Text())
	sc.Scan()
	m, _ := strconv.Atoi(sc.Text())

	recur(0, n, m)
	wr.Flush()
}

func recur(k, n, m int) {
	if k == m {
		for _, element := range selected {
			fmt.Fprintf(wr, "%d ", element)
		}
		fmt.Fprintln(wr)
	} else {
		for i := 1; i <= n; i++ {
			selected = append(selected, i)
			recur(k+1, n, m)
			selected = append(selected[:k], selected[k+1:]...)
		}
	}
}
